src/components/bulk-add/BulkAddPage.js

```diff
@@ -0,0 +1,172 @@
+import React, { useState } from 'react';
+import ScannerInput from './ScannerInput';
+import documentClient from '../../configuredDocumentClient';
+import styles from './BulkAddPage.module.css';
+
+async function getBookData(isbn) {
+    const searchParams = new URLSearchParams();
+    searchParams.set('bibkeys', `ISBN:${isbn}`);
+    searchParams.set('jscmd', 'data');
+    searchParams.set('format', 'json');
+    const response = await fetch(`/api/books?${searchParams}`, {
+        headers: {
+            'accept': 'application/json'
+        }
+    });
+    if (response.ok) {
+        const body = await response.json();
+        return body[`ISBN:${isbn}`];
+    }
+    if (response.status === 404) {
+        return '';
+    }
+    throw new Error('Error: ' + response.status);
+}
+
+function createEmptyBookWithShelf(shelf) {
+    return {
+        neckOfTheWoods: 'Library',
+        isAvailable: true,
+        bookId: '',
+        isbn: '',
+        title: '',
+        shelf
+    }
+}
+
+function isComplete({bookId, isbn, title, shelf}) {
+    return bookId && isbn && title && shelf;
+}
+
+export default function BulkAddPage() {
+    const [book, setBook] = useState(createEmptyBookWithShelf(''));
+    const [submitted, setSubmitted] = useState([]);
+
+    function onFieldChange(e) {
+        const key = e.target.name;
+        const value = e.target.value
+        setBook((previousState) => ({
+            ...previousState,
+            [key]: value
+        }));
+    }
+
+    function submitIfComplete(bookToSubmit) {
+        if (!isComplete(bookToSubmit)) {
+            return bookToSubmit;
+        }
+
+        documentClient.put({
+            TableName: process.env.REACT_APP_BOOK_TABLE,
+            Item: bookToSubmit
+        }).promise().then(() => {
+            setSubmitted((previouslySubmitted) => [bookToSubmit, ...previouslySubmitted]);
+        }, (err) => console.error(err));
+
+        return createEmptyBookWithShelf(book.shelf);
+    }
+
+    function onIdScanned(bookId) {
+        setBook((previousState) => {
+            const newState = {
+                ...previousState,
+                bookId
+            };
+            return submitIfComplete(newState);
+        });
+    }
+
+    async function onIsbnScanned(isbn) {
+        setBook((previousState) => {
+            return {
+                ...previousState,
+                isbn
+            };
+        });
+
+        const bookData = await getBookData(isbn);
+        if (bookData) {
+            setBook((previousState) => {
+                const newState = {
+                    ...previousState,
+                    title: bookData.title,
```

Comment:
When running locally, I could only get the title to populate on the first ISBN addition. Additionally, we may want to check if the length is correct before kicking off an API request. I ran into a few issues locally because I was manually typing ISBN numbers

Reply:
the ScannerInput component only fires the callback if the ISBN is the correct length I believe.

Reply:
Ah, so it does. Approved :+1: